contracts/truefi2/CrvBaseRateOracle.sol

```diff
@@ -0,0 +1,149 @@
+        curve = _curve;
+        cooldownTime = _cooldownTime;
+
+        // fill one field up of the historical buffer
+        // so the first calculateAverageRate call won't return 0
+        histBuffer.baseRates[bufferSize() - 1] = curve.get_virtual_price();
+        histBuffer.timestamps[bufferSize() - 1] = block.timestamp;
+    }
+
+    function bufferSize() public pure virtual returns (uint16) {
+        return 30;
+    }
+
+    /**
+     * @dev Helper function to get contents of the historical buffer
+     */
+    function getHistBuffer() public view returns (uint256[MAX_BUFFER_SIZE] memory, uint256[MAX_BUFFER_SIZE] memory, uint16) {
+        return (histBuffer.baseRates, histBuffer.timestamps, histBuffer.insertIndex);
+    }
+
+    /**
+     * @dev Update the historical buffer:
+     * overwrites the oldest value with current one
+     * and updates its timestamp
+     */
+    function updateRate() public offCooldown {
+        uint16 iidx = histBuffer.insertIndex;
+        histBuffer.timestamps[iidx] = block.timestamp;
+        histBuffer.baseRates[iidx] = curve.get_virtual_price();
+        histBuffer.insertIndex = uint16(iidx.add(1) % bufferSize());
+    }
+
+    /**
+     * @dev Average rate is calculated by taking
+     * the time-weighted average of the curve virtual prices.
+     * Formula is given below:
+     *
+     *           (v + v_{n-1}) / 2 + sum_{i=1}^{n - 1} (v_i + v_{i-1}) / 2 * (t_i - t_{i-1})
+     * avgRate = ---------------------------------------------------------------------------
+     *                                           (t - t_0)
+     *
+     * where v_i, t_i are values of the prices and their respective timestamps
+     * stored in the historical buffer, v is a value of current price ant t is its timestamp.
+     * Index n-1 corresponds to the most recent values and index 0 to the oldest ones.
+     * Notice that whether we are going to use the whole buffer or not
+     * depends on value of timeToCover parameter.
+     * @param timeToCover For how much time average should be calculated
+     * @return Average rate in basis points
+     */
+    function calculateAverageRate(uint256 timeToCover) public view returns (uint256) {
+        require(
+            1 days <= timeToCover && timeToCover <= 365 days,
+            "CrvBaseRateOracle: Expected amount of time in range 1 to 365 days"
+        );
+        // estimate how much buffer we need to use
+        uint256 bufferSizeNeeded = timeToCover.div(cooldownTime);
+        require(
+            bufferSizeNeeded <= bufferSize(),
+            "CrvBaseRateOracle: Needed buffer size cannot exceed size limit"
+        );
+        uint16 iidx = histBuffer.insertIndex;
+        uint256 sum;
+        uint256 totalTime;
+        for (uint16 i = 1; i < bufferSizeNeeded; i++) {
+            uint16 prevIdx = uint16(iidx.add(bufferSize()).sub(i).sub(1) % bufferSize());
+            if (histBuffer.timestamps[prevIdx] == 0)
```

Comment:
Always use {} in ifs/fors etc, we don't want to repeat [Apple's](https://blog.codecentric.de/en/2014/02/curly-braces/) mistakes